src/mapping/recurrences/grids.jl

```diff
@@ -0,0 +1,169 @@
+orginally coarse grid (a tuple of `AbstractRange`s).
+
+This approach is designed for _continuous time_ systems in which different areas of
+the state space flow may have significantly different velocity. In case of
+originally coarse grids, this may lead [`AttractorsViaRecurrences`](@ref)
+being stuck in some state space regions with
+a small motion speed and false identification of attractors. To prevent this from
+happening we provide an algorithm expansion to dynamically evaluate different regions speed
+of motion to handle areas of the grid which should be more coarse or dense than others.
+
+To achieve this, function make use of `make_irregular_array` which automatically constructs
+an array of discretization levels indices for a `grid`
+(a tuple of `AbstractRange`s) originally specified by user.
+Upon construction function automatically stores necessary parameters
+to further adopt mapping of initial conditions to specific grid density levels.
+"""
+function subdivision_based_grid(ds::DynamicalSystem, grid; maxlevel = 4)
+    lvl_array = make_irregular_array(ds, grid, maxlevel)
+    return SubdivisionBasedGrid(grid, lvl_array)
+end
+
+function SubdivisionBasedGrid(grid::NTuple{D, <:AbstractRange}, lvl_array::Array{Int, D}) where {D}
+    unique_lvls = unique(lvl_array)
+    any(<(0), unique_lvls) && error("Level array cannot contain negative values!")
+    grid_steps = Dict{Int, Vector{Int}}()
+    for i in unique_lvls
+        grid_steps[i] = [length(axis)*2^i for axis in grid]
+    end
+    grid_maxima = SVector{D, Float64}(maximum.(grid))
+    grid_minima = SVector{D, Float64}(minimum.(grid))
+
+    function scale_axis(axis, multiplier)
+        new_length = length(axis) * (2^multiplier)
+        return range(first(axis), last(axis), length=new_length)
+    end
+    multiplier = maximum(keys(grid_steps))
+    scaled_axis = [scale_axis(axis, multiplier) for axis in grid]
+    max_grid = Tuple(scaled_axis)
+
+    return SubdivisionBasedGrid(grid_steps, grid_minima, grid_maxima, lvl_array, grid, max_grid)
+end
+
+function make_irregular_array(ds::DynamicalSystem, grid, maxlevel = 4)
+    isdiscretetime(ds) && error("Dynamical system must be continuous time.")
+    indices = CartesianIndices(length.(grid))
+    f, p = dynamic_rule(ds), current_parameters(ds)
+    udummy = copy(current_state(ds))
+    velocities = zeros(length.(grid))
+    for ind in indices
+        u0 = Attractors.generate_ic_on_grid(grid, ind)
+        velocity = if !isinplace(ds)
+            f(u0, p, 0.0)
+        else
+            f(udummy, u0, p, 0.0)
+            udummy
+        end
+        if (isequal(norm(velocity),NaN))
+            velocities[ind] = Inf
+        else
+            velocities[ind] = norm(velocity)
+        end
+    end
+
+    maxvel = maximum(filter(x -> x != Inf, velocities))
+    velratios = maxvel./velocities
+    result = [round(Int,log2(clamp(x, 1, 2^maxlevel))) for x in velratios]
```

Comment:
This procedure is documented somewhere? What is the criterion for the subdivision? Why the normalization of the velocities is upside down? I would have expected `velratios = velocities/mavel`